### PR TITLE
Fix major performance regression in legacy struct marshalling APIs

### LINKED ISSUE #45100: .NET 5.0+ Marshal.PtrToStructure  is slower 8x than netcore3.1
Here is my test code:

```csharp
public class PtrBenchmarkTest
    {
        private readonly byte[] buff = new byte[Marshal.SizeOf<StructObject>()];

        public PtrBenchmarkTest()
        {
            var obj = new StructObject();
            obj.Symbol = "rb2101";
            obj.Exchange = 12;
            obj.Px = 54.90F;
            unsafe
            {
                fixed (byte* bp = buff)
                {
                    Marshal.StructureToPtr(obj, (IntPtr)bp, false);
                }
            }
        }

        [Benchmark]
        public StructObject PtrToStructure()
        {
            unsafe
            {
                fixed (byte* bp = buff)
                {
                    return (StructObject)Marshal.PtrToStructure((IntPtr)bp, typeof(StructObject));
                }
            }
        }
    }

    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public class StructObject
    {
        [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 16)]
        public string Symbol;

        [MarshalAs(UnmanagedType.I8)]
        public Int64 Exchange;

        [MarshalAs(UnmanagedType.R4)]
        public float Px;
    };

```
netcore3.1 Benchmark test result:
|         Method |      Mean |     Error |    StdDev |    Median |
|--------------- |----------:|----------:|----------:|----------:|
| PtrToStructure | 522.44 ns | 29.418 ns | 84.878 ns | 503.07 ns |

net5.0 Benchmark test result:
|         Method |        Mean |      Error |     StdDev |      Median |
|--------------- |------------:|-----------:|-----------:|------------:|
| PtrToStructure | 4,098.30 ns | 132.015 ns | 372.350 ns | 4,004.22 ns |

os: windows 10

### PR DESCRIPTION
**Pull request by jkotas**
This change adds a simple cache for the marshalling stubs. This cache improves performance of the test mentioned in the issue by 20+x.

Fixes #45100

### CODE REVIEW

**Review by jkoritzinsky — APPROVED** [review]

File: src/coreclr/vm/dllimport.cpp [hunk 1]
```diff
@@ -5373,6 +5381,8 @@ MethodDesc* NDirect::CreateStructMarshalILStub(MethodTable* pMT)
         szMetaSig.SuppressRelease();
     }
 
+    pMarshallingData->CacheStructILStub(pMT, pStubMD);
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/dllimport.cpp:5384]:
```suggestion
    // The CreateInteropILStub() handles only creating a single stub.
    // The stub returned will be okay to return even if the call below loses
    // the race to insert into the cache.
    pMarshallingData->CacheStructILStub(pMT, pStubMD);
```

**Review by AaronRobinsonMSFT — APPROVED** [review]

File: src/coreclr/vm/dllimport.cpp [hunk 2]
```diff
@@ -5282,6 +5282,14 @@ MethodDesc* NDirect::CreateStructMarshalILStub(MethodTable* pMT)
     }
     CONTRACT_END;
 
+    LoaderAllocator* pLoaderAllocator = pMT->GetLoaderAllocator();
+
+    EEMarshalingData* pMarshallingData = pLoaderAllocator->GetMarshalingData();
+
+    MethodDesc* pCachedStubMD = pMarshallingData->LookupStructILStub(pMT);
+    if (pCachedStubMD != NULL)
+        return pCachedStubMD;
```
Comment (jkotas) [inline src/coreclr/vm/dllimport.cpp:5291]:
```suggestion
        RETURN pCachedStubMD;
```

**Comment by Symbai** [comment]
Will this fix be added to .NET 6 or do we have to wait for .NET8 ?

**Comment by jkotas** [comment]
I am sorry backporting the fix to .NET 6 would not meet the servicing bug bar.

**Comment by Symbai** [comment]
I thought it might be as its a regression and a major one. Will wait for .NET 8 then.

**Pull request merged**